### PR TITLE
feat(application-templates-party-letter): Users companies field 

### PR DESCRIPTION
**Pull request by stefanhar**
# \<Description\>

We need to be able to assign a party letter to a national id.
A party letter can be assigned to a person or a company.
We have to make sure the applicant has the rights to assign a party letter to a given national id.

## What

- Added API endpoint for Iceland Revenue and Customs in central API
- Added client for Iceland Revenue and Customs API
- Added data provider for Revenue and Customs
- Added data provider for National Registry
- Added validation for selected national id
- Moved validation error builder into core
- Iterated frontend components
- Added README files
- Updated types
- Added tests

## Why

- We need to validate that the applicant has the rights to manage selected company
- We need to be able to find companies associated with applicant

## Screenshots / Gifs

Screenshot 2021-03-11 at 11 30 33

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/rsk/src/lib/rsk.spec.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
+import { rest } from 'msw'
```
Comment (jeremybarbet) [inline libs/rsk/src/lib/rsk.spec.ts:1]: I believe, according to [Eiki's comment](https://islandis.slack.com/archives/C0134BF9WG6/p1615291957054200), this client should under `libs/clients/rsk`. We would need to move `vmst` client to follow the folder convention in another PR

Comment (stefanhar) [reply libs/rsk/src/lib/rsk.spec.ts]: Wasn't sure how this was to be implemented I moved the `rsk` client into a folder `libs/clients`

File: SUMMARY.md [hunk 2]
```diff
@@ -129,6 +130,7 @@
   - [UI Shell](libs/application/ui-shell/README.md)
 - [Auth API Lib](libs/auth-api-lib/README.md)
 - [Auth Nest Tools](libs/auth-nest-tools/README.md)
+  - [Rsk](libs/clients/rsk/README.md)
```
Comment (jeremybarbet) [inline SUMMARY.md:133]: you will need to create a README at the root for `lib/clients` so gitbook will be able to create the root navigation https://docs.devland.is/handbook/repository/gitbook-contribute#root-readme-md-withing-handbook-directory

**Review by jeremybarbet — CHANGES_REQUESTED** [review]

File: SUMMARY.md (hunk 2 above)
Comment (stefanhar) [reply SUMMARY.md:134]: Fixed

File: apps/api/src/app/environments/environment.ts [hunk 3]
```diff
@@ -73,4 +73,9 @@ export default {
     username: process.env.SYSLUMENN_USERNAME ?? '',
     password: process.env.SYSLUMENN_PASSWORD ?? '',
   },
+  rskDomain: {
+    username: 'rf_api_island.is',
```
Comment (jeremybarbet) [inline apps/api/src/app/environments/environment.ts:77]: No variable needed for the username?

File: libs/api/domains/rsk/src/lib/api-domains-rsk.resolver.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import { RSKService } from '@island.is/clients/rsk'
+import { CurrentUserCompanies } from './models/currentUserCompanies.model'
+
+@UseGuards(IdsAuthGuard, ScopesGuard)
+@Resolver()
+export class RSKResolver {
+  constructor(private RSKService: RSKService) {}
+
+  @Query(() => [CurrentUserCompanies])
+  async rskGetCurrentUserCompanies(@CurrentUser() user: User) {
```
Comment (jeremybarbet) [inline libs/api/domains/rsk/src/lib/api-domains-rsk.resolver.ts:19]: Kinda picky from me, but I believe you won't use the `get` keyword to follow our graphql naming convention. Just straight `rskCurrentUserCompanies`

https://docs.devland.is/handbook/technical-overview/api-design-guide/graphql-naming-conventions#example

File: libs/api/domains/rsk/src/lib/models/currentUserCompanies.model.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { Field, ObjectType } from '@nestjs/graphql'
+
+@ObjectType()
+export class CurrentUserCompanies {
+  @Field()
+  kennitala!: string
```
Comment (jeremybarbet) [inline libs/api/domains/rsk/src/lib/models/currentUserCompanies.model.ts:6]: Translate the fields to English?

File: libs/application/templates/party-letter/src/fields/PartyLetter/index.tsx [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            label={formatMessage(m.selectPartyLetter.partyLetterLabel)}
+            placeholder={formatMessage(
+              m.selectPartyLetter.partyLetterPlaceholder,
+            )}
+            name={PARTYLETTER_ID}
+            defaultValue={
+              getValueViaPath(application.answers, PARTYLETTER_ID) as string
+            }
+            onChange={(e) => setValue(PARTYLETTER_ID, e.target.value)}
+          ></InputController>
```
Comment (jeremybarbet) [inline libs/application/templates/party-letter/src/fields/PartyLetter/index.tsx:63]: can be closed on the same opening `/>`

File: libs/application/templates/party-letter/src/fields/PartyLetter/index.tsx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            label={formatMessage(m.selectPartyLetter.partyNameLabel)}
+            placeholder={formatMessage(
+              m.selectPartyLetter.partyNamePlaceholder,
+            )}
+            name={PARTYNAME_ID}
+            defaultValue={
+              getValueViaPath(application.answers, PARTYNAME_ID) as string
+            }
+            onChange={(e) => setValue(PARTYNAME_ID, e.target.value)}
+          ></InputController>
```
Comment (jeremybarbet) [inline libs/application/templates/party-letter/src/fields/PartyLetter/index.tsx:77]: ditto

File: libs/application/templates/party-letter/src/fields/Recommendations/CopyLink.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
+import React, { FC } from 'react'
+import { FieldDescription } from '@island.is/shared/form-fields'
+import { Box, Button, Text } from '@island.is/island-ui/core'
+
+// Todo look into if this exists as util somewhere..
+const copyToClipboard = (str: string) => {
```
Comment (jeremybarbet) [inline libs/application/templates/party-letter/src/fields/Recommendations/CopyLink.tsx:6]: There is this quite popular https://www.npmjs.com/package/copy-to-clipboard but same logic as your code

File: libs/application/templates/party-letter/src/fields/Recommendations/index.tsx [hunk 9]
```diff
@@ -0,0 +1,104 @@
+import React, { FC, useState } from 'react'
+import { FieldBaseProps } from '@island.is/application/core'
+import { Box, Text, Input, Checkbox } from '@island.is/island-ui/core'
+import CopyLink from './CopyLink'
+import RecommendationTable from './RecommendationTable'
+
+const SIGNATURES = [
```
Comment (jeremybarbet) [inline libs/application/templates/party-letter/src/fields/Recommendations/index.tsx:7]: do you want to keep that here?

File: libs/application/templates/party-letter/src/lib/answerValidators.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import {
+  Application,
+  AnswerValidator,
+  AnswerValidationError,
+} from '@island.is/application/core'
+import isNumber from 'lodash/isNumber'
+import { UserCompany } from '../dataProviders/CurrentUserCompanies'
+
+const buildValidationError = (
```
Comment (jeremybarbet) [inline libs/application/templates/party-letter/src/lib/answerValidators.ts:9]: Could be moved inside the core to be reused across templates

File: apps/api/src/app/environments/environment.ts (hunk 3 above)
Comment (stefanhar) [reply apps/api/src/app/environments/environment.ts:77]: This user is already being exposed in plaintext in other parts of the project so opted for setting it here to reduce variable count in dev environment.

File: libs/api/domains/rsk/src/lib/api-domains-rsk.resolver.ts (hunk 4 above)
Comment (stefanhar) [reply libs/api/domains/rsk/src/lib/api-domains-rsk.resolver.ts:19]: Ty, will update this to match naming convention

File: libs/api/domains/rsk/src/lib/models/currentUserCompanies.model.ts (hunk 5 above)
Comment (stefanhar) [reply libs/api/domains/rsk/src/lib/models/currentUserCompanies.model.ts:6]: Yes, not sure why I didn't do this automatically 😅

File: libs/application/templates/party-letter/src/fields/PartyLetter/index.tsx (hunk 6 above)
Comment (stefanhar) [reply libs/application/templates/party-letter/src/fields/PartyLetter/index.tsx:63]: Fixed

File: libs/application/templates/party-letter/src/fields/Recommendations/index.tsx (hunk 9 above)
Comment (stefanhar) [reply libs/application/templates/party-letter/src/fields/Recommendations/index.tsx:7]: This is a WIP and will be populated with correct data via another PR, this is just here to display correct UI

File: libs/application/templates/party-letter/src/lib/answerValidators.ts (hunk 10 above)
Comment (stefanhar) [reply libs/application/templates/party-letter/src/lib/answerValidators.ts:9]: Nicked this from parental-leave, will move into core and update other templates using the same solution

File: libs/application/templates/party-letter/src/fields/Recommendations/CopyLink.tsx (hunk 8 above)
Comment (stefanhar) [reply libs/application/templates/party-letter/src/fields/Recommendations/CopyLink.tsx:6]: Seems to have more compatibility will swap this out

File: libs/api/domains/rsk/src/lib/models/currentUserCompanies.model.ts [hunk 11]
```diff
@@ -3,17 +3,20 @@ import { Field, ObjectType } from '@nestjs/graphql'
...
-  stadaAdila!: string
+  operationalForm!: string
 
   @Field()
-  erStjorn!: '0' | '1'
+  companyStatus!: string
 
   @Field()
-  erProkuruhafi!: '0' | '1'
+  isPartOfBoardOfDirectors!: '0' | '1'
```
Comment (jeremybarbet) [inline libs/api/domains/rsk/src/lib/models/currentUserCompanies.model.ts:18]: Boolean?

**Review by jeremybarbet — COMMENTED** [review]
One more question, after that it LGTM 👍

File: libs/api/domains/rsk/src/lib/models/currentUserCompanies.model.ts (hunk 11 above)
Comment (stefanhar) [reply libs/api/domains/rsk/src/lib/models/currentUserCompanies.model.ts]: Yes, fixed

**Review by jeremybarbet — APPROVED** [review]
LGTM 💯

**Review by siggakogk — APPROVED** [review]

**Review by siggakogk — APPROVED** [review]

**Review by dd-sdang — APPROVED** [review]

**Pull request merged**